orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/core/graph/pipeline_transformer.h"
+
+using namespace onnxruntime::common;
+
+namespace onnxruntime {
+namespace training {
+
+void GetPipelineSendOutput(const Graph& graph, std::string& loss_name) {
+  for (auto& node : graph.Nodes()) {
+    if (!node.OpType().compare("Send")) {
+      // send op should always have an output, which is the OutputSignal.
+      loss_name = node.OutputDefs()[0]->Name();
+      return;
+    }
+  }
+}
+
+Status TransformGraphForPipeline(Graph& graph) {
+  // insert WaitEvent and RecordEvent to the partition
+  Node* send_fw{nullptr};
+  Node* send_bw{nullptr};
+  Node* recv_fw{nullptr};
+  Node* recv_bw{nullptr};
+  for (auto& node : graph.Nodes()) {
+    if (node.OpType() == "Send") {
+      if (node.Description() == "Backward pass") {
+        send_bw = &node;
+      } else {
+        send_fw = &node;
+      }
+    } else if (node.OpType() == "Recv") {
+      if (node.Description() == "Backward pass") {
+        recv_bw = &node;
+      } else {
+        recv_fw = &node;
+      }
+    }
+  }
+
+  ORT_RETURN_IF_NOT(
+          (send_fw && recv_bw) || (!send_fw && !recv_bw),
```

Comment:
>(send_fw && recv_bw) || (!send_fw && !recv_bw) [](start = 10, length = 46)

send_fw == recv_bw

Reply:
ah, was thinking them as bool, then maybe !!(send_fw) == !!(recv_bw)

---
In reply to: [408344271](https://github.com/microsoft/onnxruntime/pull/3513#discussion_r408344271) [](ancestors = 408344271,408340457)

Reply:
We should use more readable ones, like nullptr!=send_fw.

Reply:
This logic is more like (send_fw XNOR recv_bw). I have updated the code based on Ke's suggestion.

---
In reply to: [408378340](https://github.com/microsoft/onnxruntime/pull/3513#discussion_r408378340) [](ancestors = 408378340)